smpclient/__init__.py

```diff
@@ -97,32 +98,53 @@ async def __aenter__(self) -> 'SMPClient':
     async def __aexit__(self) -> None:
         await self.disconnect()
 
-    def _maximize_packet(self, request: ImageUploadWrite, image: bytes) -> ImageUploadWrite:
-        """Given an `ImageUploadWrite` with empty `data`, return the largest packet possible."""
+    @staticmethod
+    def _cbor_integer_size(integer: int) -> int:
+        """CBOR integers are packed as small as possible."""
+
+        return 0 if integer < 24 else 1 if integer < 0xFF else 2 if integer < 0xFFFF else 4
+
+    def _get_max_cbor_and_data_size(self, request: smpmsg.WriteRequest) -> Tuple[int, int]:
+        """Given an `ImageUploadWrite`, return the maximum CBOR size and data size."""
+
+        h: Final = cast(smpheader.Header, request.header)
+
+        # given empty data in the request, how many bytes are available for the data?
+        unencoded_bytes_available: Final = self._transport.max_unencoded_size - len(bytes(request))
 
-        def cbor_integer_size(integer: int) -> int:
-            """CBOR integers are packed as small as possible."""
-            return 0 if integer < 24 else 1 if integer < 0xFF else 2 if integer < 0xFFFF else 4
+        # how many bytes are required to encode the data size?
+        bytes_required_to_encode_data_size: Final = self._cbor_integer_size(
+            unencoded_bytes_available
+        )
+
+        # the final data size is the unencoded bytes available minus the bytes
+        # required to encode the data size
+        data_size: Final = max(0, unencoded_bytes_available - bytes_required_to_encode_data_size)
+
+        # the final CBOR size is the original header length plus the data size
+        # plus the bytes required to encode the data size
+        cbor_size: Final = h.length + data_size + self._cbor_integer_size(data_size)
 
-        _h = cast(smpheader.Header, request.header)
+        return cbor_size, data_size
+
+    def _maximize_packet(self, request: ImageUploadWrite, image: bytes) -> ImageUploadWrite:
+        """Given an `ImageUploadWrite` with empty `data`, return the largest packet possible."""
 
-        chunk_size = self._transport.max_unencoded_size - len(bytes(request))
-        chunk_size -= cbor_integer_size(chunk_size)
-        chunk_size = max(0, chunk_size)
-        cbor_size = _h.length + chunk_size + cbor_integer_size(chunk_size)
+        h: Final = cast(smpheader.Header, request.header)
```

Comment:
Curious as to why `_h` was 'private' before?

Reply:
I was using it as shorthand for "this is a temporary value" but decided it was more clear to just make it "h" for "header".